### PR TITLE
fix: allow [$ dune subst] to work without a repo

### LINKED ISSUE #10927: dune subst silently ignores %%ID%% string when outside a git-controlled directory
Note this impact file-pinned opam packages in particular
## Reproduction

```sh
$ mkdir test

$ cat > dune-project << EOF
(lang dune 3.16)
(name test)
EOF

$ echo "%%NAME%%" > README.md
```

## Expected Behavior

```sh
$ dune subst
$ cat README.md
test
```

## Actual Behavior

```sh
$ dune subst
$ cat README.md
%%NAME%%
```

## Specifications

- dune 3.16.0
- ocaml 5.2
- debian 12

### PR DESCRIPTION
**Pull request by rgrinberg**
Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

### CODE REVIEW

**Comment by rgrinberg** [comment]
@anmonteiro do you want to review this one?

File: bin/subst.ml [hunk 1]
```diff
@@ -299,111 +299,142 @@ let make_watermark_map ~commit ~version ~dune_project ~info =
...
+  (match vcs with
+   | Some vcs ->
+     let+ version = Vcs.describe vcs
+     and+ commit_id = Vcs.commit_id vcs
+     and+ files = Vcs.files vcs in
+     Some (version, commit_id, files)
+   | None ->
+     (* We have to do this because scanning the source tree evaluates [-p].
+        That's because [-p] is needed to interpret packages in dune projects
+        correctly. It should not be necessary, so we shuold probably make the
```
Comment (anmonteiro) [inline bin/subst.ml:311]:
```suggestion
        correctly. It should not be necessary, so we should probably make the
```

File: bin/subst.ml [hunk 2]
```diff
@@ -299,111 +299,142 @@ let make_watermark_map ~commit ~version ~dune_project ~info =
...
+        package loading lazier. *)
+     Dune_rules.Only_packages.Clflags.set No_restriction;
+     let* root = Source_tree.root () in
+     let project = Source_tree.Dir.project root in
+     if Dune_project.dune_version project < (3, 17)
+     then Memo.return None
```
Comment (anmonteiro) [inline bin/subst.ml:317]: should we print a warning / hint here or would that be too noisy?

File: bin/subst.ml [hunk 3]
```diff
@@ -299,111 +299,142 @@ let make_watermark_map ~commit ~version ~dune_project ~info =
...
+      then subst_file path ~map:watermarks opam_package_files))
 ;;
 
 let subst () =
-  match
-    Sys.readdir "." |> Array.to_list |> String.Set.of_list |> Vcs.Kind.of_dir_contents
-  with
-  | None -> Fiber.return ()
-  | Some kind -> Memo.run (subst { kind; root = Path.root })
+  Sys.readdir "."
```
Comment (anmonteiro) [inline bin/subst.ml:431]: why do we need `Sys.readdir "."` at all if we get the source root in the function above?

Would it be better to find the root first and then try to find the VCS kind? what happens if `subst` is not run from the VCS root?

**Review by anmonteiro — APPROVED** [review]

File: bin/subst.ml (hunk 2 above)
Comment (rgrinberg) [reply bin/subst.ml:317]: Too noisy I'd say. Essentially every single package would emit this warning now.

File: bin/subst.ml (hunk 3 above)
Comment (rgrinberg) [reply bin/subst.ml:431]: Yeah, we can call `Source_tree.nearest_vcs`. Will leave a TODO

**Comment by btjorge** [comment]
The fix introduce a regression when an opam file is present, I just opened #11045 reporting that

**Pull request merged**